File: shared/ChatHistoryStore.cs
using System.Text.Json;

namespace Shared;

/// <summary>
/// 채팅 메시지를 로컬 파일(%AppData%\MultiChat\...)에 한 줄당 JSON 하나씩 append로 저장하고,
/// 다음 실행 때 그대로 불러온다. 서버/클라이언트가 각자 자신이 본 메시지를 저장하므로 파일을 분리한다.
/// </summary>
public class ChatHistoryStore
{
    private readonly string _path;

    public ChatHistoryStore(string fileName)
    {
        var dir = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "MultiChat");
        Directory.CreateDirectory(dir);
        _path = Path.Combine(dir, fileName);
    }

    public List<ChatEnvelope> Load()
    {
        var result = new List<ChatEnvelope>();
        if (!File.Exists(_path)) return result;

        foreach (var line in File.ReadLines(_path, LineProtocol.NoBom))
        {
            if (string.IsNullOrWhiteSpace(line)) continue;
            try
            {
                var env = JsonSerializer.Deserialize<ChatEnvelope>(line);
                if (env != null) result.Add(env);
            }
            catch { /* 손상된 줄은 건너뛴다 */ }
        }
        return result;
    }

    // ponytail: 파일이 무한정 커질 수 있음 - 강의실 단기 사용 기준으로는 문제 없음. 필요해지면 날짜별 로테이션 추가.
    public void Append(ChatEnvelope env)
    {
        try
        {
            File.AppendAllText(_path, JsonSerializer.Serialize(env) + "\n", LineProtocol.NoBom);
        }
        catch { /* 저장 실패해도 채팅 자체는 계속되어야 함 */ }
    }
}

File: mac/Sources/MacChatClient/ChatHistoryStore.swift
import Foundation

/// shared/ChatHistoryStore.cs와 대응. ~/Library/Application Support/MultiChat/에 한 줄당 JSON 하나씩 저장.
final class ChatHistoryStore {
    private let url: URL

    init(fileName: String) {
        let dir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("MultiChat")
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        url = dir.appendingPathComponent(fileName)
    }

    func load() -> [ChatEnvelope] {
        guard let text = try? String(contentsOf: url, encoding: .utf8) else { return [] }
        var result: [ChatEnvelope] = []
        for line in text.split(separator: "\n") {
            if line.trimmingCharacters(in: .whitespaces).isEmpty { continue }
            if let data = line.data(using: .utf8), let env = try? JSONDecoder().decode(ChatEnvelope.self, from: data) {
                result.append(env)
            }
        }
        return result
    }

    // ponytail: 파일이 무한정 커질 수 있음 - 강의실 단기 사용 기준으로는 문제 없음. 필요해지면 날짜별 로테이션 추가.
    func append(_ env: ChatEnvelope) {
        guard let data = try? JSONEncoder().encode(env), let json = String(data: data, encoding: .utf8) else { return }
        let line = Data((json + "\n").utf8)
        if let handle = try? FileHandle(forWritingTo: url) {
            defer { try? handle.close() }
            handle.seekToEndOfFile()
            handle.write(line)
        } else {
            try? line.write(to: url)
        }
    }
}
